uni-stark/examples/prove_monty_31_keccak.rs

```diff
@@ -0,0 +1,326 @@
+use std::env;
+use std::fmt::Debug;
+
+use p3_baby_bear::{BabyBear, GenericPoseidon2LinearLayersBabyBear};
+use p3_blake3_air::Blake3Air;
+use p3_challenger::SerializingChallenger32;
+use p3_commit::ExtensionMmcs;
+use p3_dft::{Radix2DitParallel, TwoAdicSubgroupDft};
+use p3_field::extension::BinomialExtensionField;
+use p3_field::{ExtensionField, PrimeField32, TwoAdicField};
+use p3_fri::{FriConfig, TwoAdicFriPcs};
+use p3_keccak::{Keccak256Hash, KeccakF};
+use p3_keccak_air::KeccakAir;
+use p3_koala_bear::{GenericPoseidon2LinearLayersKoalaBear, KoalaBear};
+use p3_merkle_tree::MerkleTreeMmcs;
+use p3_monty_31::dft::RecursiveDft;
+use p3_poseidon2::GenericPoseidon2LinearLayers;
+use p3_poseidon2_air::{RoundConstants, VectorizedPoseidon2Air};
+use p3_symmetric::{CompressionFunctionFromHasher, PaddingFreeSponge, SerializingHasher32To64};
+use p3_uni_stark::{prove, verify, StarkConfig, SymbolicExpression};
+use rand::distributions::Standard;
+use rand::prelude::Distribution;
+use rand::{random, thread_rng};
+use tracing_forest::util::LevelFilter;
+use tracing_forest::ForestLayer;
+use tracing_subscriber::layer::SubscriberExt;
+use tracing_subscriber::util::SubscriberInitExt;
+use tracing_subscriber::{EnvFilter, Registry};
+
+const P2_WIDTH: usize = 16;
+const P2_HALF_FULL_ROUNDS: usize = 4;
+const P2_VECTOR_LEN: usize = 1 << 3;
+
+fn prove_hashes<
+    F: PrimeField32 + TwoAdicField,
+    EF: ExtensionField<F> + TwoAdicField,
+    DFT: TwoAdicSubgroupDft<F>,
+    LinearLayers: GenericPoseidon2LinearLayers<F, P2_WIDTH>
+        + GenericPoseidon2LinearLayers<SymbolicExpression<F>, P2_WIDTH>
+        + GenericPoseidon2LinearLayers<F::Packing, P2_WIDTH>
+        + GenericPoseidon2LinearLayers<EF, P2_WIDTH>,
+    const P2_SBOX_DEGREE: u64,
+    const P2_SBOX_REGISTERS: usize,
+    const P2_PARTIAL_ROUNDS: usize,
+>(
+    proof_goal: &str,
```

Comment:
It might be nice to have this take a little `enum ProofGoal { Blake3, Poseidon2, Keccak }`, partly to avoid the double-parsing, and partly just for consistency so all arguments would be structured / already parsed.